File: LeetCode/LeetCode Practice Problems/1261- Find Elements in a Contaminated Binary Tree.cpp
class FindElements {
public:
    TreeNode* r;
    map<int,bool> m;
    
    FindElements(TreeNode* root) {
        r = root;
        
        queue<TreeNode*> q1;
        queue<int> q2;
        q1.push(r);
        q2.push(0);
        
        while(!q1.empty()) {
            int val = q2.front();
            q2.pop();
            TreeNode* t = q1.front();
            q1.pop();
            
            m[val] = true;
            
            if(t->left != NULL) {
                q1.push(t->left);
                q2.push(val*2+1);
            }
            if(t->right != NULL) {
                q1.push(t->right);
                q2.push(val*2+2);
            }
        }
        
    }
    
    bool find(int target) {
        return m[target];
        
    }
};

File: LeetCode/LeetCode Practice Problems/1261-Find Elements in a Contaminated Binary Tree.java
class FindElements {
    HashSet<Integer> set = new HashSet();
    
    public void traverse(TreeNode root,int val) {
        if(root == null) return;
        set.add(val);
        traverse(root.left , 2*val+1);
        traverse(root.right , 2*val+2);
    }
    
    public FindElements(TreeNode root) {
        traverse(root,0);
    }
    
    public boolean find(int target) {
        return set.contains(target);
    }
}
